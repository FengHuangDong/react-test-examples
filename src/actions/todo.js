import axios from 'axios';
import { REMOVE_TODO, ADD_TODO, SET_VISIBILITY_FILTER,FETCH_TODOS,ERR_FETCH_TODOS, TOGGLE_TODO } from "./actionsTypes";

export const addToList = (text) => {
  return {
    type: ADD_TODO,
    payload: text
  };
};

export const setVisibilityFilter = (type) => {
  return {
    type:SET_VISIBILITY_FILTER,
    payload:type
  }
}
export const toogleTodo = (id) => {
  return {
    type:TOGGLE_TODO,
    payload:id
  }
}
export const removeFromList = (text) => {
  return {
      type:REMOVE_TODO,
      payload: text
  }
};


export const fetchTodos = () => {
  return async (dispatch) => {
    try {
      let todos = await axios.get(`https://jsonplaceholder.typicode.com/todos`)
      dispatch({
        type: FETCH_TODOS,
        payload: todos
      })
    } catch (err) {
      debugger
      dispatch({
        type: ERR_FETCH_TODOS,
        payload: err
      })
    }
  }
}
