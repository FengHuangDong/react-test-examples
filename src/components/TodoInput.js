import React, { Component } from 'react'
import { Field, reduxForm } from 'redux-form'

export class TodoInput extends Component {

  render() {
    const { handleSubmit } = this.props
    return (
      <div className="form-group row">
        <div className="col-sm-10">
          <form className="form-inline" onSubmit={ handleSubmit }>
            <Field className="form-control"  placeholder="add todo here" name="title" component="input" type="text" />
            <button style={{marginLeft:"10px"}} className="btn btn-success" type="submit">Submit</button>
          </form>
        </div>
      </div>
    )
  }
}


export default reduxForm({
  // a unique name for the form
  form: 'todo'
})(TodoInput);
