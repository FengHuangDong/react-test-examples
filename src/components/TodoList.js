import React, { Component } from 'react'
import PropTypes from 'prop-types';

import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import * as todoActions from '../actions/todo';
import { SHOW_ALL, SHOW_COMPLETED, SHOW_ACTIVE } from "../actions/actionsTypes";

export class TodoList extends Component {
  render () {
    return (
      <React.Fragment>
        <nav style={{ marginTop: "60px" }}>
          <ol className="breadcrumb">
            <li
              className={"breadcrumb-item "+ (this.props.visibilityFilter === SHOW_ALL ? 'active' : '') }
              onClick={() => this.props.actions.setVisibilityFilter(SHOW_ALL)}
            >
             All
            </li>
            <li
               className={"breadcrumb-item "+ (this.props.visibilityFilter === SHOW_COMPLETED ? 'active' : '') }
              onClick={() => this.props.actions.setVisibilityFilter(SHOW_COMPLETED)}
            >
              Completed
            </li>
            <li
               className={"breadcrumb-item "+ (this.props.visibilityFilter === SHOW_ACTIVE ? 'active' : '') }
              onClick={() => this.props.actions.setVisibilityFilter(SHOW_ACTIVE)}
            >
              Active
            </li>
          </ol>
        </nav>
        {
          this.props.todos.length !== 0 ? (
            <table
              style={{ marginTop: "60px" }}
              className="table table-hover table-dark"
            >
              <thead>
                <tr>
                  <th scope="col">Todos</th>
                  <th scope="col">Actions</th>
                </tr>
              </thead>
              <tbody>
                {this.props.todos.map((todo,i) => (
                  <tr key={i}>
                    <td
                      style={{
                        textDecoration: todo.completed ? "line-through" : "none"
                      }}
                    >
                      {todo.title} {todo.completed === true ? "(completed)" : ""}
                    </td>
                    <td>
                      <span
                        className="fas fa-minus-circle"
                        onClick={() => this.props.actions.removeFromList(todo)}
                        style={{
                          color: "white",
                          fontSize: "20pt",
                          marginRight: "20px"
                        }}
                      />
                      <span
                        className="fas fa-check-circle"
                        onClick={() => this.props.actions.toogleTodo(todo.id)}
                        style={{ color: "white", fontSize: "20pt" }}
                      />
                    </td>
                  </tr>
                ))}
              </tbody>
            </table>
          ) : (
            <div
              style={{ marginTop: "50px" }}
              className="col-lg-10 col-md-10 col-xs-12 col-sm-12 offset-lg-1"
            >
              <div className="alert alert-danger" role="alert">
                Todo List is empty or Filter results show no results
              </div>
            </div>
          )
        }
      </React.Fragment>
    )
  }
}

TodoList.propTypes = {
  todos: PropTypes.array
}
export const getVisibleTodos = (todos, filter) => {
  switch (filter) {
    case SHOW_ALL:
      return todos;
    case SHOW_COMPLETED:
      return todos.filter(t => t.completed);
    case SHOW_ACTIVE:
      return todos.filter(t => !t.completed);
    default:
      throw new Error("Unknown filter: " + filter);
  }
};

export function mapStateToProps(state, props) {
    return {
      todos: getVisibleTodos(state.todos, state.visibilityFilter),
      visibilityFilter: state.visibilityFilter
    }
}

export function mapDispatchToProps(dispatch) {
    return {
        actions: bindActionCreators(todoActions, dispatch)
    }
}


export default connect(mapStateToProps, mapDispatchToProps)(TodoList);
