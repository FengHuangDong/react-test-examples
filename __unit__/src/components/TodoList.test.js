import React from 'react';
import {TodoList, getVisibleTodos} from "../../../src/components/TodoList"
import {shallow} from 'enzyme'
import toJson from "enzyme-to-json"
import { SHOW_ALL, SHOW_COMPLETED, SHOW_ACTIVE } from "../../../src/actions/actionsTypes";

const mockTodos = [
  {
    "userId": 1,
    "id": 1,
    "title": "delectus aut autem",
    "completed": false
  },
  {
    "userId": 1,
    "id": 2,
    "title": "quis ut nam facilis et officia qui",
    "completed": false
  },
  {
    "userId": 1,
    "id": 3,
    "title": "fugiat veniam minus",
    "completed": false
  },
  {
    "userId": 1,
    "id": 4,
    "title": "et porro tempora",
    "completed": true
  }
]

describe('TodoList', () => {
  let mockProps, wrapper
  beforeEach(() => {
    mockProps = {
      actions:{
        removeFromList:jest.fn(),
        setVisibilityFilter:jest.fn(),
        toogleTodo:jest.fn()
      },
      todos:mockTodos,
      visibilityFilter:SHOW_ALL
    }
    wrapper = shallow(<TodoList {...mockProps} />)
  });
  it('should render component', () => {
    let todoList = wrapper.find('tbody tr')
    expect(todoList.length).toEqual(mockProps.todos.length)
    expect(toJson(wrapper)).toMatchSnapshot();
  });
  it('should not todo data render', () => {
    wrapper.setProps({ todos: [] });
    expect(wrapper.find(".alert-danger")).toHaveLength(1);
  });
  it('should action to be call when visibility filter click', () => {
    wrapper.find(".breadcrumb-item").map($li => {
      $li.simulate('click');
      expect(mockProps.actions.setVisibilityFilter).toHaveBeenCalled();
    })
  });
  it('should filter active when set visibilityFilter props', () => {
    wrapper.setProps({visibilityFilter: SHOW_COMPLETED})
    expect(wrapper.find(".breadcrumb-item").at(1).hasClass("active")).toEqual(true)
    wrapper.setProps({visibilityFilter: SHOW_ACTIVE})
    expect(wrapper.find(".breadcrumb-item").at(2).hasClass("active")).toEqual(true)
  })
  it('should remverFromList to be call when click icon',() => {
    wrapper.find(".fa-minus-circle").at(0).simulate("click")
    expect(mockProps.actions.removeFromList).toHaveBeenCalled();
  })
  it('should toogleTodo to be call when click icon',() => {
    wrapper.find(".fa-check-circle").at(0).simulate("click")
    expect(mockProps.actions.toogleTodo).toHaveBeenCalled();
  })

});

describe('getVisibleTodos filter', () => {
  it('should SHOW_ALL todos', () => {
    expect(getVisibleTodos(mockTodos,SHOW_ALL)).toEqual(mockTodos);
  });
  it('should SHOW_COMPLETED todos', () => {
    expect(getVisibleTodos(mockTodos,SHOW_COMPLETED)).toEqual([{
      "userId": 1,
      "id": 4,
      "title": "et porro tempora",
      "completed": true
    }]);
  });
  it('should SHOW_ACTIVE todos', () => {
    expect(getVisibleTodos(mockTodos,SHOW_ACTIVE)).toEqual([
      {
        "userId": 1,
        "id": 1,
        "title": "delectus aut autem",
        "completed": false
      },
      {
        "userId": 1,
        "id": 2,
        "title": "quis ut nam facilis et officia qui",
        "completed": false
      },
      {
        "userId": 1,
        "id": 3,
        "title": "fugiat veniam minus",
        "completed": false
      }]
    );
  });
  it('should error', () => {
    expect(()=>{getVisibleTodos(mockTodos,'test')}).toThrow();
  });
});