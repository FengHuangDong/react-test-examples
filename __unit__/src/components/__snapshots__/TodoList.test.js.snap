// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList should render component 1`] = `
<Fragment>
  <nav
    style={
      Object {
        "marginTop": "60px",
      }
    }
  >
    <ol
      className="breadcrumb"
    >
      <li
        className="breadcrumb-item active"
        onClick={[Function]}
      >
        All
      </li>
      <li
        className="breadcrumb-item "
        onClick={[Function]}
      >
        Completed
      </li>
      <li
        className="breadcrumb-item "
        onClick={[Function]}
      >
        Active
      </li>
    </ol>
  </nav>
  <table
    className="table table-hover table-dark"
    style={
      Object {
        "marginTop": "60px",
      }
    }
  >
    <thead>
      <tr>
        <th
          scope="col"
        >
          Todos
        </th>
        <th
          scope="col"
        >
          Actions
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        key="0"
      >
        <td
          style={
            Object {
              "textDecoration": "none",
            }
          }
        >
          delectus aut autem
           
        </td>
        <td>
          <span
            className="fas fa-minus-circle"
            onClick={[Function]}
            style={
              Object {
                "color": "white",
                "fontSize": "20pt",
                "marginRight": "20px",
              }
            }
          />
          <span
            className="fas fa-check-circle"
            onClick={[Function]}
            style={
              Object {
                "color": "white",
                "fontSize": "20pt",
              }
            }
          />
        </td>
      </tr>
      <tr
        key="1"
      >
        <td
          style={
            Object {
              "textDecoration": "none",
            }
          }
        >
          quis ut nam facilis et officia qui
           
        </td>
        <td>
          <span
            className="fas fa-minus-circle"
            onClick={[Function]}
            style={
              Object {
                "color": "white",
                "fontSize": "20pt",
                "marginRight": "20px",
              }
            }
          />
          <span
            className="fas fa-check-circle"
            onClick={[Function]}
            style={
              Object {
                "color": "white",
                "fontSize": "20pt",
              }
            }
          />
        </td>
      </tr>
      <tr
        key="2"
      >
        <td
          style={
            Object {
              "textDecoration": "none",
            }
          }
        >
          fugiat veniam minus
           
        </td>
        <td>
          <span
            className="fas fa-minus-circle"
            onClick={[Function]}
            style={
              Object {
                "color": "white",
                "fontSize": "20pt",
                "marginRight": "20px",
              }
            }
          />
          <span
            className="fas fa-check-circle"
            onClick={[Function]}
            style={
              Object {
                "color": "white",
                "fontSize": "20pt",
              }
            }
          />
        </td>
      </tr>
      <tr
        key="3"
      >
        <td
          style={
            Object {
              "textDecoration": "line-through",
            }
          }
        >
          et porro tempora
           
          (completed)
        </td>
        <td>
          <span
            className="fas fa-minus-circle"
            onClick={[Function]}
            style={
              Object {
                "color": "white",
                "fontSize": "20pt",
                "marginRight": "20px",
              }
            }
          />
          <span
            className="fas fa-check-circle"
            onClick={[Function]}
            style={
              Object {
                "color": "white",
                "fontSize": "20pt",
              }
            }
          />
        </td>
      </tr>
    </tbody>
  </table>
</Fragment>
`;
